jena-arq/src/main/java/org/apache/jena/sparql/engine/binding/BindingBuilder.java

```diff
@@ -0,0 +1,486 @@
+        set$(var, node);
+        return this;
+    }
+
+    // Accumulate (var,value) pairs.
+    public BindingBuilder add(Var var, Node node) {
+        checkAdd(var, node, false);
+        add$(var, node);
+        return this;
+    }
+
+    /** Add all the (var, value) pairs from another binding */
+    public BindingBuilder addAll(Binding other) {
+        Iterator<Var> vIter = other.vars();
+        for (; vIter.hasNext(); ) {
+            Var v = vIter.next();
+            Node n = other.get(v);
+            add(v,n);
+        }
+        return this;
+    }
+
+    private void add$(Var var, Node node)  {
+        if ( haveBuilt )
+            throw new IllegalStateException("Attempt to add a binding pair after build() without reset()");
+
+        if ( map != null ) {
+            map.put(var, node);
+            return;
+        }
+        if ( var1 == null ) {
+            var1 = var;
+            node1 = node;
+            return;
+        }
+        if ( var2 == null ) {
+            var2 = var;
+            node2 = node;
+            return;
+        }
+        if ( var3 == null ) {
+            var3 = var;
+            node3 = node;
+            return;
+        }
+        if ( var4 == null ) {
+            var4 = var;
+            node4 = node;
+            return;
+        }
+        // Spill from 4 to N. map is null.
+        map = new HashMap<>();
+        map.put(var1, node1);
+        map.put(var2, node2);
+        map.put(var3, node3);
+        map.put(var4, node4);
+        map.put(var, node);
+        var1 = null;
+        node1 = null;
+        var2 = null;
+        node2 = null;
+        var3 = null;
+        node3 = null;
+        var4 = null;
+        node4 = null;
+    }
+
+    private void set$(Var var, Node node)  {
+        if ( haveBuilt )
+            throw new IllegalStateException("Attempt to set a binding pair after build() without reset()");
+
+        if ( map != null ) {
+            map.put(var, node);
+            return;
+        }
+        if ( var1 == null || var.equals(var1) ) {
+            var1 = var;
+            node1 = node;
+            return;
+        }
+        if ( var2 == null || var.equals(var2) ) {
+            var2 = var;
+            node2 = node;
+            return;
+        }
+        if ( var3 == null || var.equals(var3) ) {
+            var3 = var;
+            node3 = node;
+            return;
+        }
+        if ( var4 == null || var.equals(var4) ) {
+            var4 = var;
+            node4 = node;
+            return;
+        }
+        // Spill from 4 to N.
+        // map is null.
+        map = new HashMap<>();
+        map.put(var1, node1);
+        map.put(var2, node2);
+        map.put(var3, node3);
+        map.put(var4, node4);
+        var1 = null;
+        node1 = null;
+        var2 = null;
+        node2 = null;
+        var3 = null;
+        node3 = null;
+        var4 = null;
+        node4 = null;
+        map.put(var, node);
```

Comment:
Took me a few seconds re-reading it to see that `$add` and `$set` are not identical :smile: need more :coffee:

Reply:
... or maybe less :coffee: 😄